### PR TITLE
Add docs for ASN1_item_sign and ASN1_item_verify functions

### PR DESCRIPTION
**Pull request by slontis**
This is to address part of issue #13192.

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
The travis failure seems very strange - but unrelated to this PR.

**Comment by slontis** [comment]
Yes that is a bit weird..

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man3/ASN1_item_sign.pod [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+ int ASN1_item_verify_ctx(const ASN1_ITEM *it, const X509_ALGOR *alg,
+                          const ASN1_BIT_STRING *signature, const void *data,
+                          EVP_MD_CTX *ctx);
+
+=head1 DESCRIPTION
+
+ASN1_item_sign_ex() is a helper function that is used internally by other
+functions such as X509_sign() to sign an object (such as a certificate) using
+the ASN.1 structure I<it>, algorithm identifiers I<algor1> and I<algor2>,
```
[the remark this answers is not in the document]
Comment (slontis) [reply doc/man3/ASN1_item_sign.pod]: I was really hoping you would say that :)
Updated.

**Comment by levitte** [comment]
Quite frankly, I'd like to throw these functions as far away as I can.  They are 100% geared for legacy, the entanglement with EVP_PKEY_ASN1_METHOD ctrl functions runs deep...

File: doc/man3/ASN1_item_sign.pod [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                      EVP_PKEY *pkey);
+
+ int ASN1_item_verify_ctx(const ASN1_ITEM *it, const X509_ALGOR *alg,
+                          const ASN1_BIT_STRING *signature, const void *data,
+                          EVP_MD_CTX *ctx);
+
+=head1 DESCRIPTION
+
+ASN1_item_sign_ex() is used to sign arbitrary ASN1 data using a data object
+I<data>, the ASN.1 structure I<it>, algorithm identifiers I<algor1> and I<algor2>,
```
Comment (mattcaswell) [inline doc/man3/ASN1_item_sign.pod:43]: What do algor1 and algor2 actually *do*?

File: doc/man3/ASN1_item_sign.pod [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+ASN1_ITEM_rptr(X509_CINF), ASN1_ITEM_rptr(X509_REQ_INFO) and
+ASN1_ITEM_rptr(X509_CRL_INFO).
+The B<OSSL_LIB_CTX> specified in I<libctx> and the property query string
+specified in I<props> are used when searching for algorithms in providers.
+The generated signature is set into I<signature>.
+The optional parameter I<id> can be NULL, but can be set for special key types.
+See EVP_PKEY_CTX_set1_id() for further info. If I<algor1> or I<algor2> are not
+NULL then they are set to the AlgorithmIdentifier that corresponds to the
+combination of signature algorithm and digest algorithm for the signature
+operation of the I<pkey>.
```
Comment (mattcaswell) [inline doc/man3/ASN1_item_sign.pod:57]: Oh..so these are output parameters??? Can we make that clearer?

File: doc/man3/ASN1_item_sign.pod [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+       * cert->cert_info is the X509_CINF object that will be
+       * converted into DER data and then signed.
+       * cert->cert_info.signature is filled with the private key's signing
+       * algorithm id.
+       * cert->sig_alg is another copy of the signing algorithm id that sits
+       * within the cert object (it must match the cert->cert_info.signature)
+       * when verifying.
+       */
+      return ASN1_item_sign_ctx(it,
+                                &cert->cert_info.signature, &cert->sig_alg,
```
Comment (mattcaswell) [inline doc/man3/ASN1_item_sign.pod:110]: X509 is an opaque type so this won't work for external uses.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man3/ASN1_item_sign.pod (hunk 2 above)
Comment (levitte) [reply doc/man3/ASN1_item_sign.pod:43]: That has never been well explained, and I've seen use cases where they are used very differently.  Even looking through git history, it remains a mystery, at least as far as I was able to dig last time I tried to figure this out.

In X.509 terms, it looks like `algor1` is meant the get the algorithm from `signature` from `TBScertificate`, and `algor2` is meant to get the algorithm from `signatureAlgorithm` in `Certificate`.
(I still have a hard time understanding the difference between those two algos, so can't help further)

However, with other uses, those two get other algorithms.  As I recall, PKCS7 and CMS use these, with their own assumptions.

In the end, this depends hugely on the use, and the called of `ASN1_item_sign_ex()` *must* know the backend that does the grunt of the work intimately, and it all comes down to which ctrl commands are used.

File: doc/man3/ASN1_item_sign.pod (hunk 4 above)
Comment (slontis) [reply doc/man3/ASN1_item_sign.pod]: Ok I have made another attempt at this now by creating a custom object in the example.
Not sure that it is any easier to understand.

File: doc/man3/ASN1_item_sign.pod (hunk 2 above)
Comment (kroeckx) [reply doc/man3/ASN1_item_sign.pod:43]: They are output parameters, not input. The to be signed data, and the struct with the signature can both have copy of the X509_ALGOR. The ASN1_item_sign() function will write the correct X509_ALGOR to the structs based on the algorithms and parameters that have been set up. If one of the algors points to the X509_ALGOR of the TBS data, that X509_ALGOR will first be written before the signature is generated.

File: doc/man3/ASN1_item_sign.pod [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+      * obj->info is the 'MySignObject' object that will be
+      * converted into DER data and then signed.
+      * obj->signature will contain the output signature.
+      * obj->sig_alg is filled with the private key's signing algorithm id.
+      * obj->info.sig_alg is another copy of the signing algorithm id that sits
+      * within the MyObject object.
+      */
+     return ASN1_item_sign_ctx(it, &obj->sig_alg, &obj->info.sig_alg,
+                               &obj->signature, &obj->info, mdctx);
+ }
```
Comment (kroeckx) [inline doc/man3/ASN1_item_sign.pod:141]: The memset() isn't making much sense to me, you're removing the data from obj->info.

I wonder if it's useful to also created an ASN1 object for MyObject and then did i2d on that to convert the whole thing to ASN1.

Comment (kroeckx) [reply doc/man3/ASN1_item_sign.pod:141]: The example should probably also cover verify

Comment (slontis) [reply doc/man3/ASN1_item_sign.pod:203]: Thanks for the input.. That was a useful suggestion.
See what you think of the example now.

**Comment by slontis** [comment]
ping

File: doc/man3/ASN1_item_sign.pod [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+converting it to der format using the ASN.1 structure I<it>.
+Valid values that can be used by the ASN.1 structure I<it> are
+ASN1_ITEM_rptr(X509_CINF), ASN1_ITEM_rptr(X509_REQ_INFO) and
+ASN1_ITEM_rptr(X509_CRL_INFO).
+The B<OSSL_LIB_CTX> specified in I<libctx> and the property query string
+specified in I<props> are used when searching for algorithms in providers.
+The generated signature is set into I<signature>.
+The optional parameter I<id> can be NULL, but can be set for special key types.
+See EVP_PKEY_CTX_set1_id() for further info. The output parameters <algor1> and
+I<algor2> are set to the AlgorithmIdentifier that corresponds to the
```
Comment (t8m) [inline doc/man3/ASN1_item_sign.pod:55]: From this it is not clear what is the difference between algor1 and algor2 or why do we even have two parameters.

Comment (slontis) [reply doc/man3/ASN1_item_sign.pod:55]: Any suggestions?..
As you can see it is not easy to describe..
It allows you to optionally put the algorithm into an info object you are going to sign as well as a cached copy in a higher level object that contains the info object. From the point of view of this function it just allows you to set it twice...

@kroeckx said this above..
The to be signed data, and the struct with the signature can both have copy of the X509_ALGOR. The ASN1_item_sign() function will write the correct X509_ALGOR to the structs based on the algorithms and parameters that have been set up. If one of the algors points to the X509_ALGOR of the TBS data, that X509_ALGOR will first be written before the signature is generated.

Comment (t8m) [reply doc/man3/ASN1_item_sign.pod:55]: Can you just use what @kroeckx said?

Comment (slontis) [reply doc/man3/ASN1_item_sign.pod]: updated with a variant of the above.

**Review by t8m — APPROVED** [review]

**Comment by slontis** [comment]
Thanks. Merged to master.

**Pull request closed without merging**